pkg/scheduler/algorithm/predicates/predicates.go

```diff
@@ -1277,74 +1279,80 @@ func getMatchingTopologyPairs(pod *v1.Pod, nodeInfoMap map[string]*schedulercach
 			catchError(fmt.Errorf("node not found"))
 			return
 		}
-		nodeTopologyMaps := newTopologyPairsMaps()
 		for _, existingPod := range nodeInfo.PodsWithAffinity() {
-			affinity := existingPod.Spec.Affinity
-			if affinity == nil {
-				continue
-			}
-			for _, term := range GetPodAntiAffinityTerms(affinity.PodAntiAffinity) {
-				namespaces := priorityutil.GetNamespacesFromPodAffinityTerm(existingPod, &term)
-				selector, err := metav1.LabelSelectorAsSelector(term.LabelSelector)
-				if err != nil {
-					catchError(err)
-					return
-				}
-				if priorityutil.PodMatchesTermsNamespaceAndSelector(pod, namespaces, selector) {
-					if topologyValue, ok := node.Labels[term.TopologyKey]; ok {
-						pair := topologyPair{key: term.TopologyKey, value: topologyValue}
-						nodeTopologyMaps.addTopologyPair(pair, existingPod)
-					}
-				}
+			existingPodTopologyMaps, err := getMatchingTopologyPairsOfExistingPod(pod, existingPod, node)
+			if err != nil {
+				catchError(err)
+				return
 			}
-			if len(nodeTopologyMaps.podToTopologyPairs) > 0 {
-				appendTopologyPairsMaps(nodeTopologyMaps)
+
+			if len(existingPodTopologyMaps.podToTopologyPairs) > 0 {
+				appendTopologyPairsMaps(existingPodTopologyMaps)
 			}
 		}
 	}
 	workqueue.Parallelize(16, len(allNodeNames), processNode)
 	return topologyMaps, firstError
 }
 
+// getMatchingTopologyPairsOfExistingPod is a helper to calculate on each existing pod
+// of given node, whether (1) it has PodAntiAffinity and (2) AffinityTerms matches incoming pod
 func getMatchingTopologyPairsOfExistingPod(newPod *v1.Pod, existingPod *v1.Pod, node *v1.Node) (*topologyPairsMaps, error) {
 	topologyMaps := newTopologyPairsMaps()
 	affinity := existingPod.Spec.Affinity
-	if affinity != nil && affinity.PodAntiAffinity != nil {
-		for _, term := range GetPodAntiAffinityTerms(affinity.PodAntiAffinity) {
-			namespaces := priorityutil.GetNamespacesFromPodAffinityTerm(existingPod, &term)
-			selector, err := metav1.LabelSelectorAsSelector(term.LabelSelector)
-			if err != nil {
-				return nil, err
-			}
-			if priorityutil.PodMatchesTermsNamespaceAndSelector(newPod, namespaces, selector) {
-				if topologyValue, ok := node.Labels[term.TopologyKey]; ok {
-					pair := topologyPair{key: term.TopologyKey, value: topologyValue}
-					topologyMaps.addTopologyPair(pair, existingPod)
-				}
-			}
+	if affinity == nil || affinity.PodAntiAffinity == nil {
+		return topologyMaps, nil
+	}
+
+	matchMap := make(map[string]bool) // key is TopologyKey
```

Comment:
It seems we only put comment above the line of code :-)

Reply:
Having topologyKey as a key for this map (matchMap) could lead to an undesired behaviour as terms could have different topology keys, and accordingly the bug would still occur in case if terms with other topology keys are not matching it will still be added to the final result. https://github.com/kubernetes/kubernetes/blob/0d0d5bf0e4b312f375cad2d0a36e07a8b1baf711/pkg/scheduler/algorithm/predicates/predicates.go#L1330-L1333 Am I correct?

Reply:
@ahmad-diaa It's a good point - terms have different TopologyKey in the same Pod{Anti}Affinity. And there is no testcase covering this. Let me think about which is the right behavior.

Reply:
@ahmad-diaa is this scenario what you are concerned?

```
- we have 1 zone called z1
- nodeA and nodeB is in z1
- there is an existing pod having:
    - nodeName: "nodeA"
    - anti-affinity terms:
        - {label "foo" exists, topologyKey as "invalid-node-label"}
        - {label "bar" exists, topologyKey as "zone"}
- a pod comes in:
    - labels: {"foo": "", "bar": ""}
```

With and Without fix in this PR, incoming pod won't fit on any node - i.e. Pending. But actually I _think_ correct behavior is it should be able to fit on any node. Because no node has label "invalid-node-label", so the entire anti-affinity terms of exiting pod is invalid.

^^ @bsalamat your opinion on above scenario?

Reply:
@ahmad-diaa BTW, in addition to above "topologyKey-not-exist" case, I don't see other cases current PR has problems. If you got some examples, please share here.

Reply:
@ahmad-diaa this might be another case you're concerned:

```
- we have 1 region called r1
- in r1, we have 2 zones called z1 and z2
- nodeA and nodeB are located in r1/z1 and r1/z2 respectively
- there is an existing pod having:
    - nodeName: "nodeA"
    - anti-affinity terms:
        - {label "foo" exists, topologyKey as "region"}
        - {label "bar" exists, topologyKey as "zone"}
- a pod comes in:
    - labels: {"foo": "", "bar": ""}
```

The result turns out to be:
- with this fix, incoming pod is "pending"
- without this fix but on latest master branch (say included your recent performance PR), incoming pod is also "pending"
- in 1.11.0, incoming pod landed on nodeB

cc @bsalamat to clarify which behavior is correct.

Reply:
@Huang-Wei In my opinion, the topologyKey value should not affect the decision at this point  as it is checked later in "satisifiesExistingPodAntiAffinity" method: https://github.com/kubernetes/kubernetes/blob/0d0d5bf0e4b312f375cad2d0a36e07a8b1baf711/pkg/scheduler/algorithm/predicates/predicates.go#L1389-L1394
However, this still checks for only one term match which is not the behaviour we are looking for, so the above lines can be replaced with:
https://github.com/kubernetes/kubernetes/blob/0d0d5bf0e4b312f375cad2d0a36e07a8b1baf711/pkg/scheduler/algorithm/predicates/predicates.go#L1406

Also, I think "targetPodMatchesAntiAffinityOfPod(existingPod, pod)" could be used to decide whether or not to add the pod's terms to topologyPairMaps
https://github.com/kubernetes/kubernetes/blob/0d0d5bf0e4b312f375cad2d0a36e07a8b1baf711/pkg/scheduler/algorithm/predicates/metadata.go#L449
Using the above mentioned method would enhance readability, and it will be consistent with checking the terms of the pod being scheduled.

So, if the approach mentioned above is used, the incoming pod (in the scenario described) would fit because the node labels won't contain invalid-node-label and "nodeMatchesTopologyTerms" will return false.

Finally, I have another concerning scenario:
```
- we have 1 zone called z1 and 1 region called r1
- nodeA in zone z2 and region r1
- nodeB in zone z2 and region r2
- there are two existing pods:
  -"pod1" having:
     - nodeName: "nodeA"
     - anti-affinity terms:
         - {label "foo" exists, topologyKey as "zone"}
         - {label "bar" exists, topologyKey as "region"}
  -"pod2" having:
     - nodeName: "nodeB"
     - anti-affinity terms:
         - {label "bar" exists, topologyKey as "region"}
- a pod comes in to be scheduled on nodeB:
    - labels: {"foo": "", "bar": ""}
```
The incoming pod would be considered no fit as all it's terms are matched but that is done through different pods (which should not be the case, as far as I understand). So, the following line should be updated to check that all matches are from the same pod:
https://github.com/kubernetes/kubernetes/blob/0d0d5bf0e4b312f375cad2d0a36e07a8b1baf711/pkg/scheduler/algorithm/predicates/predicates.go#L1411

Reply:
> However, this still checks for only one term match which is not the behaviour we are looking for, so the above lines can be replaced with:

Agree.

> So, if the approach mentioned above is used, the incoming pod (in the scenario described) would fit because the node labels won't contain invalid-node-label and "nodeMatchesTopologyTerms" will return false.

Yeap, that's the desired behavior I think.

> Finally, I have another concerning scenario:

Could you revise this example a little bit... Some typos there I think, and region is a bigger concept than zone (regions has multiple zones) :)

Reply:
After reading thru the context, I think it's clear (and seems we've reached conclusion) that "matching of all items" should be considered with same topologyKey.

> Finally, I have another concerning scenario:

It seems the point in this case is: when two existing Pods spread on two Nodes in same topology domain, the coming Pod with `{foo, bar}` should match all items for every Pod, not `foo` from one Pod and `bar` from the other.

Reply:
@Huang-Wei your understanding and examples are correct.

@ahmad-diaa In the example that you gave:
```
- nodeA in r1/z1
- nodeB in r1/z2 
- there are two existing pods:
  -"pod1" having:
     - nodeName: "nodeA"
     - anti-affinity terms:
         - {label "foo" exists, topologyKey as "zone"}
  -"pod2" having:
     - nodeName: "nodeB"
     - anti-affinity terms:
         - {label "bar" exists, topologyKey as "region"}
- a pod comes in to be scheduled on nodeA:
    - labels: {"foo": "", "bar": ""}
```

The incoming pod does not have an assigned node when created. It should not be scheduled, as it violates anti-affinity rule of Pod1 and shouldn't be scheduled to node B as it violates the anti-affinity rule of Pod2.